pkg/core/native/notary.go

```diff
@@ -0,0 +1,358 @@
+	n.AddMethod(md, desc, false)
+
+	desc = newDescriptor("postPersist", smartcontract.VoidType)
+	md = newMethodAndPrice(getOnPersistWrapper(postPersistBase), 0, smartcontract.AllowModifyStates)
+	n.AddMethod(md, desc, false)
+
+	return n
+}
+
+// Metadata implements Contract interface.
+func (n *Notary) Metadata() *interop.ContractMD {
+	return &n.ContractMD
+}
+
+// Initialize initializes Notary native contract and implements Contract interface.
+func (n *Notary) Initialize(ic *interop.Context) error {
+	return nil
+}
+
+// OnPersist implements Contract interface.
+func (n *Notary) OnPersist(ic *interop.Context) error {
+	var (
+		nFees    int64
+		notaries keys.PublicKeys
+		err      error
+	)
+	for _, tx := range ic.Block.Transactions {
+		if tx.HasAttribute(transaction.NotaryAssistedT) {
+			if notaries == nil {
+				notaries, err = n.GetNotaryNodes(ic.DAO)
+				if err != nil {
+					return fmt.Errorf("failed to get notary nodes: %w", err)
+				}
+			}
+			nKeys := tx.GetAttributes(transaction.NotaryAssistedT)[0].Value.(*transaction.NotaryAssisted).NKeys
+			nFees += int64(nKeys) + 1
+			if tx.Sender() == n.Hash {
+				payer := tx.Signers[1]
+				balance := n.getDepositFor(ic.DAO, payer.Account)
+				balance.Amount.Sub(balance.Amount, big.NewInt(tx.SystemFee+tx.NetworkFee))
+				if balance.Amount.Sign() == 0 {
+					err := n.removeDepositFor(ic.DAO, payer.Account)
+					if err != nil {
+						return fmt.Errorf("failed to remove an empty deposit for %s from storage: %w", payer.Account.StringBE(), err)
+					}
+				} else {
+					err := n.putDepositFor(ic.DAO, balance, payer.Account)
+					if err != nil {
+						return fmt.Errorf("failed to update deposit for %s: %w", payer.Account.StringBE(), err)
+					}
+				}
+			}
+		}
+	}
+	if nFees == 0 {
+		return nil
+	}
+	singleReward := calculateNotaryReward(nFees, len(notaries))
+	for _, notary := range notaries {
+		n.GAS.mint(ic, notary.GetScriptHash(), singleReward)
+	}
+	return nil
+}
+
+// onPayment records deposited amount as belonging to "from" address with a lock
+// till the specified chain's height.
+func (n *Notary) onPayment(ic *interop.Context, args []stackitem.Item) stackitem.Item {
+	if h := ic.VM.GetCallingScriptHash(); h != n.GAS.Hash {
+		panic(fmt.Errorf("only GAS can be accepted for deposit, got %s", h.StringBE()))
+	}
+	from := toUint160(args[0])
+	to := from
+	amount := toBigInt(args[1])
+	data, ok := args[2].(*stackitem.Array)
+	if !ok || len(data.Value().([]stackitem.Item)) != 2 {
+		panic(errors.New("invalid data parameter"))
+	}
+	additionalParams := data.Value().([]stackitem.Item)
+	if bytes, err := additionalParams[0].TryBytes(); err == nil {
+		to, err = util.Uint160DecodeBytesBE(bytes)
+		if err != nil {
+			panic(fmt.Errorf("failed to decode `to` from `data` parameter: %w", err))
+		}
+	}
+	bigInt, err := additionalParams[1].TryInteger()
+	if err != nil {
+		panic(fmt.Errorf("failed to decode `till` from `data` parameter: %w", err))
+	}
+	till := uint32(bigInt.Int64())
```

Comment:
Better check it first for `IsInt64()` and then for fitting into `uint32` as it's invalid otherwise.